From ea249ebe6a7acf4426de35935973f841f05cb268 Mon Sep 17 00:00:00 2001
From: Alistair Buxton <a.j.buxton@gmail.com>
Date: Thu, 1 Sep 2011 15:48:35 +0100
Subject: [PATCH 2/6] Store existing atags at startup if chainloading.

This patch causes u-boot to store any Linux kernel atags which it was passed
by a previous bootloader. The original load address of u-boot is also stored.
The data is stored at 0x82000000.

Signed-off-by: Alistair Buxton <a.j.buxton@gmail.com>
---
 arch/arm/cpu/armv7/start.S |   14 ++++++++++++++
 1 files changed, 14 insertions(+), 0 deletions(-)

diff --git a/arch/arm/cpu/armv7/start.S b/arch/arm/cpu/armv7/start.S
index a41ecce..4e2df58 100644
--- a/arch/arm/cpu/armv7/start.S
+++ b/arch/arm/cpu/armv7/start.S
@@ -120,6 +120,20 @@ IRQ_STACK_START_IN:
 
 reset:
 #if (CONFIG_CHAINLOADER)
+	/*
+	 * Ugly hack: store the u-boot load address and kernel params from
+	 * previous bootloader.
+	 * 0x82000000: address where this image was loaded
+	 * 0x82000004: mach_type
+	 * 0x82000008: atags address
+	 */
+	adr 	r0, _start		@ Where u-boot.bin was loaded
+	mov 	r3, #0x82000000		@ Dest addr. for saved data
+	str	r0, [r3]
+	add	r3, r3, #4
+	str	r1, [r3]
+	add	r3, r3, #4
+	str	r2, [r3]
 	/* Copy u-boot.bin to TEXT_BASE if it isn't already there. */
 	adr	r0, _start		@ Where u-boot.bin was loaded
 	adr	r2, _start
-- 
1.7.6.1

